### PR TITLE
Add `rank_zero_first` utility

### LINKED ISSUE #17733: Utility to let rank 0 do work while other ranks wait
### Description & Motivation

We often see this pattern, for example when downloading data on rank 0 only:

```py
if fabric.global_rank == 0:
    # download data on rank 0 only

fabric.barrier()  # all processes sync here
```

The barrier is important so processes don't fall out of sync, but it is easy to forget.

### Pitch

Provide a utility (function or method) context manager that combines both the rank check and the barrier:

```py
with fabric.rank_zero_only:
   # download data on rank 0 only

# all processes sync here
```
The barrier would simply reside in the `__exit__` of the context manager. A similar one could be for `node_zero_only`.

### Alternatives

Not do it. Leave the code as is. Con: Error prone if one forgets the barrier.

### Additional context

Motivated by workarounds like #17732

Full example based on the linked issue:

```py
import lightning as L
import torch; import torchvision as tv

fabric = L.Fabric()
fabric.launch()

model = tv.models.resnet18()
optimizer = torch.optim.SGD(model.parameters(), lr=0.001)
model, optimizer = fabric.setup(model, optimizer)

with fabric.rank_zero_only:
    # download data
	tv.datasets.CIFAR10("data", download=True, train=True, transform=tv.transforms.ToTensor())

dataset = tv.datasets.CIFAR10("data", download=False, train=True, transform=tv.transforms.ToTensor())

dataloader = torch.utils.data.DataLoader(dataset, batch_size=8)
dataloader = fabric.setup_dataloaders(dataloader)

model.train()
num_epochs = 10
for epoch in range(num_epochs):
    for batch in dataloader:
        inputs, labels = batch
        optimizer.zero_grad()
        outputs = model(inputs)
        loss = torch.nn.functional.cross_entropy(outputs, labels)
        fabric.backward(loss)
        optimizer.step()
        print(loss.data)
```

cc @borda @awaelchli @carmocca @justusschock

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #17733 

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

cc @borda @carmocca @justusschock @awaelchli

### CODE REVIEW

**Review by Borda — APPROVED** [review]

**Review by carmocca — APPROVED** [review]
Cool!
